idl/partition_checksum.idl.hh

```diff
@@ -13,6 +13,7 @@
 #include "idl/uuid.idl.hh"
 #include "idl/frozen_mutation.idl.hh"
 #include "idl/token.idl.hh"
+#include "repair/id.hh"
```

Comment:
Is the include still needed?

Reply:
it used to be `repair/repair.hh` which is too high level and kinda reversed the dependencies.
With moving the id defs into this low-level include file, both repair and idl depend on it which is nicer.

Reply:
we still have a reversal, no? idl depends on repair, and repair depends on idl.

Reply:
idl doesn't depend on repair per-se, only  on repair id. what's the problem of depending on the low level definition?

Reply:
> it used to be repair/repair.hh which is too high level and kinda reversed the dependencies.
With moving the id defs into this low-level include file, both repair and idl depend on it which is nicer.

Benny, idl/partition_checksum.idl.hh never included repair/repair.hh. 

I think it is wrong to include repair/id.hh,  idl/partition_checksum.idl.hh. We'd better keep the runtime structure and the on wire idl definition separately, even if the idl has a very similar syntax.  

We should modify idl very carefully, including the runtime repair/id.hh can accidentally break idl.

Reply:
> > it used to be repair/repair.hh which is too high level and kinda reversed the dependencies.
> > With moving the id defs into this low-level include file, both repair and idl depend on it which is nicer.
> 
> Benny, idl/partition_checksum.idl.hh never included repair/repair.hh.
> 
> I think it is wrong to include repair/id.hh, idl/partition_checksum.idl.hh. We'd better keep the runtime structure and the on wire idl definition separately, even if the idl has a very similar syntax.
> 
> We should modify idl very carefully, including the runtime repair/id.hh can accidentally break idl.

I disagree. IDL should break by design if the classes it depends on change, not by accident.
This was the motivation behind 1fda686f966ebe1ccb39c5c0e1e310184d80a3fa

Reply:
Say we have foo.hh.

Before:
foo {
uint64 x;
};

After:
foo {
uint64 x;
uint64 _some_internal_data;
};

if we include foo.hh directly in idl things will break.

if we have 

foo.idl

foo {
uint64 x
};

Things will not break.

You see the difference.

Reply:
If the c++ code depends on the idl classes, it shouldn't redefine them, but rather include the generated idl.hh header file.
But if the idl structures depend on c++ structures they should include the respective c++ headers to resolve the dependencies, otherwise the c++ headers need to be particularly included by the c++ code that includes the idl header file - which was really messy before 1fda686f966ebe1ccb39c5c0e1e310184d80a3fa 

In your example above I don't understand the order of dependency.

Reply:
> If the c++ code depends on the idl classes, it shouldn't redefine them, but rather include the generated idl.hh header file. 

In what way a c++ code may depend on idl classes? IMO this is a wrong way to use the IDL. IDL should describe how to encode a certain data type on wire which may be different from how the object is represented in c++.

Reply:
It's not that the idl definition file depends on the c++ header, but rather the idl generated files may depend on c++ definitions.

For example see https://github.com/scylladb/scylladb/pull/11673#discussion_r998097704

Cc @avikivity